sdk/tables/Azure.Data.Tables/src/TableBatchResponse.cs

```diff
@@ -32,12 +31,46 @@ internal TableBatchResponse(ConcurrentDictionary<string, (HttpMessage Message, R
         /// <returns></returns>
         public Response GetResponseForEntity(string rowKey)
         {
-            if (!_requestLookup.TryGetValue(rowKey, out (HttpMessage Message, RequestType RequestType) tuple))
+            if (!_requestLookup.TryGetValue(rowKey, out HttpMessage message))
             {
                 throw new InvalidOperationException("The batch operation did not contain an entity with the specified rowKey");
             }
 
-            return tuple.Message.Response;
+            return message.Response;
+        }
+
+        /// <summary>
+        /// Tries to get the entity that caused the batch operation failure from the <see cref="RequestFailedException"/>.
+        /// </summary>
+        /// <param name="exception">The exception thrown from <see cref="TableClient.SubmitTransaction"/> or <see cref="TableClient.SubmitTransactionAsync"/>.</param>
+        /// <param name="submittedBatchItems">The submitted list of <see cref="BatchItem"/>. This list should be unchanged since it was submitted via
+        /// <see cref="TableClient.SubmitTransaction"/> or <see cref="TableClient.SubmitTransactionAsync"/> </param>
+        /// <param name="failedEntity">If the return value is <c>true</c>, contains the <see cref="ITableEntity"/> that caused the batch operation to fail.</param>
+        /// <returns><c>true</c> if the failed entity was retrieved from the exception, else <c>false</c>.</returns>
+        public static bool TryGetFailedEntityFromException(RequestFailedException exception, List<BatchItem> submittedBatchItems, out ITableEntity failedEntity)
+        {
+            Argument.AssertNotNull(exception, nameof(exception));
+            Argument.AssertNotNull(submittedBatchItems, nameof(submittedBatchItems));
+
+            failedEntity = null;
+
+            if (!exception.Data.Contains(TableConstants.ExceptionData.FailedEntityIndex))
+            {
+                return false;
+            }
+            try
+            {
+                if (exception.Data[TableConstants.ExceptionData.FailedEntityIndex] is string stringIndex && int.TryParse(stringIndex, out int index))
+                {
+                    failedEntity = submittedBatchItems[index].Entity;
+                }
+            }
+            catch
+            {
+                // We just don't want to throw here.
```

Comment:
Should we consider returning `false` if an exception occurs?

Reply:
This is going to change a bit based on arch board feedback - so this concern will go away.